spec/validator.ml

```diff
@@ -0,0 +1,868 @@
+      (_ : nonces_commitments IntMap.t) : Secp256k1.scalar * string =
+    failwith "place holder"
+end
+
+module Abi = struct
+  let encode (_ : 'a list) (_ : 'b) : string = failwith "place holder"
+  let encode_packed (_ : 'a list) (_ : 'b) : string = failwith "place holder"
+
+  let encode_with_selector (_ : 'a) (_ : 'b list) (_ : 'c) : string =
+    failwith "place holder"
+
+  let hash_typed_data (_ : 'a) : string = failwith "place holder"
+end
+
+module MerkleTree = struct
+  let root (_ : 'a) : string = failwith "place holder"
+  let proof (_ : 'a) (_ : 'b) : string list = failwith "place holder"
+end
+
+module ParticipantSet = struct
+  type t = unit (* place holder *)
+
+  let init (_ : Address.t list) : t = failwith "place holder"
+  let remove (_ : t) (_ : 'a) : t = failwith "place holder"
+  let remove_missing (_ : t) (_ : 'a) : t = failwith "place holder"
+  let cardinal (_ : t) : int = failwith "place holder"
+  let identifier (_ : t) (_ : Address.t) : int = failwith "place holder"
+  let addr (_ : t) (_ : int) : Address.t = failwith "place holder"
+end
+
+type group_id = string
+
+module FROSTCoordinator = struct
+  let group_id (_ : ParticipantSet.t) (_ : int) (_ : int) (_ : string) :
+      group_id =
+    failwith "place holder"
+end
+
+type group = {
+  id : group_id;  (** The set of participants in the group. *)
+  participants : ParticipantSet.t;
+      (** The identifier of the validator in the group. *)
+  key : Secp256k1.public_key;  (** The participant share ID. *)
+  share : Secp256k1.secret_key;
+}
+
+type epoch = { id : int; group : group }
+
+type rollover =
+  | Collecting_key_gen_commitments of {
+      epoch : int;
+      id : group_id;
+      coefficients : Secp256k1.scalar list;
+      commitments : Secp256k1.point list IntMap.t;
+      participants : ParticipantSet.t;
+      deadline : int;
+    }
+  | Collecting_key_gen_secret_shares of {
+      epoch : int;
+      id : group_id;
+      key : Secp256k1.point;
+      commitments : Secp256k1.point list IntMap.t;
+      shares : Secp256k1.scalar IntMap.t;
+      last_participant : Address.t;
+      participants : ParticipantSet.t;
+      deadline : int;
+    }
+  | Signing_epoch_rollover of { epoch : epoch; message : string }
+  | Staged_epoch of epoch
+
+type preprocess_group = {
+  nonces : FROST.nonces IntMap.t;
+  pending : FROST.nonces list StringMap.t;
+}
+
+type signature_id = string
+
+type sign_status =
+  | Waiting_for_sign_request of { responsible : Address.t }
+  | Collecting_sign_nonces of {
+      id : signature_id;
+      nonces : FROST.nonces_commitments IntMap.t;
+    }
+  | Collecting_sign_shares of {
+      id : signature_id;
+      root : string;
+      group_commitment : Secp256k1.point;
+      shares : Secp256k1.scalar IntMap.t;
+    }
+  | Waiting_for_consensus_attestation of {
+      id : signature_id;
+      responsible : Address.t;
+    }
+
+type packet =
+  | Epoch_rollover of { epoch : int; rollover : int; group_id : group_id }
```

Comment:
this is not how the packet is defined onchain, as there is no "offchain" way to link from a groupKey to a groupId

Reply:
Good point